File: app/src/main/java/com/soundscore/app/ui/viewmodel/ListsViewModel.kt
package com.soundscore.app.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.soundscore.app.data.model.UserList
import com.soundscore.app.data.repository.AppContainer
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

data class ListsUiState(
    val lists: List<UserList> = emptyList(),
    val showcases: List<ListShowcase> = emptyList(),
    val syncMessage: String? = null,
)

class ListsViewModel : ViewModel() {
    private val repository = AppContainer.repository

    val uiState: StateFlow<ListsUiState> = combine(
        repository.lists,
        repository.albums,
        repository.syncMessage,
    ) { lists, albums, syncMessage ->
        ListsUiState(
            lists = lists,
            showcases = resolveListShowcases(lists, albums),
            syncMessage = syncMessage,
        )
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5_000),
        initialValue = ListsUiState(),
    )

    fun createList(title: String) {
        viewModelScope.launch {
            repository.createList(title)
        }
    }
}

File: ios/SoundScore/SoundScore/ViewModels/ListsViewModel.swift
import Foundation
import Combine

class ListsViewModel: ObservableObject {
    @Published var lists: [UserList]
    @Published var showcases: [ListShowcase]
    @Published var syncMessage: String?
    @Published var isLoading: Bool
    @Published var errorMessage: String?

    init() {
        let repo = SoundScoreRepository.shared
        self.lists = repo.lists
        self.showcases = resolveListShowcases(repo.lists, repo.albums)
        self.syncMessage = repo.syncMessage
        self.isLoading = repo.isLoading
        self.errorMessage = repo.errorMessage

        repo.$lists
            .receive(on: RunLoop.main)
            .assign(to: &$lists)

        Publishers.CombineLatest(repo.$lists, repo.$albums)
            .receive(on: RunLoop.main)
            .map { resolveListShowcases($0, $1) }
            .assign(to: &$showcases)

        repo.$syncMessage
            .receive(on: RunLoop.main)
            .assign(to: &$syncMessage)

        repo.$isLoading
            .receive(on: RunLoop.main)
            .assign(to: &$isLoading)

        repo.$errorMessage
            .receive(on: RunLoop.main)
            .assign(to: &$errorMessage)
    }

    func createList(title: String) {
        SoundScoreRepository.shared.createList(title: title)
    }
}
